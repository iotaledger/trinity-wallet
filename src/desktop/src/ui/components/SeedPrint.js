import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import QRCode from 'qr.js/lib/QRCode';

import { byteToChar } from 'libs/crypto';

import paperWallet from 'themes/paper-wallet.svg';
import paperWalletFilled from 'themes/paper-wallet-filled.svg';

import css from './seedPrint.scss';

const wallets = {
    paperWallet: paperWallet,
    paperWalletFilled: paperWalletFilled,
};

export default class SeedPrint extends PureComponent {
    static propTypes = {
        /** Should print filled wallet */
        filled: PropTypes.bool.isRequired,
        /** Seed byte array */
        seed: PropTypes.array.isRequired,
        /** Seed checksum */
        checksum: PropTypes.string.isRequired,
    };

    render() {
        const { filled, seed, checksum } = this.props;

        const qr = new QRCode(-1, 1);

        seed.forEach((byte) => {
            qr.addData(byteToChar(byte));
        });
        qr.make();

        const cells = qr.modules;

        return (
            <div className={css.print} onClick={() => window.print()}>
                {filled && (
                    <svg viewBox="0 0 595 841" xmlns="http://www.w3.org/2000/svg">
                        {seed.map((byte, index) => {
                            const letter = byteToChar(byte % 27);
                            const space = index % 9 > 5 ? 38 : index % 9 > 2 ? 19 : 0;
                            const x = 193 + (index % 9) * 26 + space;
                            const y = 365 + Math.floor(index / 9) * 32.8;
                            return (
                                <text x={x} y={y} key={`${index}${letter}`}>
                                    {letter}
                                </text>
                            );
                        })}
                        {cells.map((row, rowIndex) => {
                            return row.map((cell, cellIndex) => (
                                <rect
                                    height={1.6}
                                    key={cellIndex}
                                    style={{ fill: cell ? '#000000' : 'none' }}
                                    width={1.6}
                                    x={160 + cellIndex * 1.6}
                                    y={698 + rowIndex * 1.6}
                                />
                            ));
                        })}
                        <text x="278" y="735">
                            {checksum}
                        </text>
                    </svg>
                )}
                <img width="auto" height="100vh" src={filled ? wallets.paperWallet : wallets.paperWalletFilled} />
            </div>
        );
    }
}
