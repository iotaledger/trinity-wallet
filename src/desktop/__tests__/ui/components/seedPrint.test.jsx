import React from 'react';
import { shallow } from 'enzyme';

import SeedPrint from 'ui/components/SeedPrint';

const props = {
    filled: false,
    seed: Array(81).fill(0),
    checksum: 'ABC',
};

describe('SeedPrint component', () => {
    test('Render the component', () => {
        const wrapper = shallow(<SeedPrint {...props} />);

        expect(wrapper).toMatchSnapshot();
    });

    test('Empty template', () => {
        const wrapper = shallow(<SeedPrint {...props} />);

        expect(wrapper.find('svg')).toHaveLength(0);
    });

    test('Filled template', () => {
        const mockProps = Object.assign({}, props, { filled: true });
        const wrapper = shallow(<SeedPrint {...mockProps} />);

        expect(wrapper.find('path')).toHaveLength(mockProps.seed.length + mockProps.checksum.length);
    });
});
