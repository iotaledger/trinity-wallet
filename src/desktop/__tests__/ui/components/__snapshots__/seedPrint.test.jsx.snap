// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeedPrint component Render the component 1`] = `
<div
  className="print"
  onClick={[Function]}
>
  <img
    height="100vh"
    src={Object {}}
    width="auto"
  />
</div>
`;
